### PR TITLE
x64: Improve memory support in `{insert,extract}lane`

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit improves adds support to Cranelift to emit `pextr{b,w,d,q}` with a memory destination, merging a store-of-extract operation into one instruction. Additionally AVX support is added for the `pextr*` instructions.

I've additionally tried to ensure that codegen tests and runtests exist for all forms of these instructions too.

### CODE REVIEW

**Comment by alexcrichton** [comment]
This is a generalization of https://github.com/bytecodealliance/wasmtime/pull/5924 for more vector types and more lanes.

File: cranelift/codegen/src/isa/x64/lower.isle [hunk 1]
```diff
@@ -2659,17 +2659,41 @@
...
-(rule 3 (lower (store flags
-                    (has_type (ty_64 _) (extractlane value (u8_from_uimm8 0)))
+(rule 2 (lower (store flags
+                    (has_type $F64 (extractlane value (u8_from_uimm8 0)))
                     address
                     offset))
       (side_effect
        (x64_movsd_store (to_amode flags address offset) value)))
+(rule 2 (lower (store flags
+                    (has_type $I8 (extractlane value (u8_from_uimm8 n)))
```
Comment (abrown) [inline cranelift/codegen/src/isa/x64/lower.isle:2674]: I checked and `u8_from_uimm8` does not guarantee that `n` is in the right range (e.g., 0 to 15 here). Did you see any validation that would prevent wrong `n` here at the CLIF level or do we rely exclusively on the Wasm-level construction? If there is nothing and it ends up being too tricky to do here in ISLE, maybe `emit.rs` could gain some `assert!`s or something like that.

File: cranelift/filetests/filetests/isa/x64/extractlane-avx.clif [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+block0(v0: f64x2):
+  v1 = extractlane v0, 1
+  return v1
+}
+
+; VCode:
+;   pushq   %rbp
+;   movq    %rsp, %rbp
+; block0:
+;   vpshufd $238, %xmm0, %xmm0
```
Comment (abrown) [inline cranelift/filetests/filetests/isa/x64/extractlane-avx.clif:139]: This is `0b11101110`... is that right?

File: cranelift/filetests/filetests/isa/x64/insertlane-avx.clif [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+  v3 = insertlane v0, v2, 0
+  return v3
+}
+
+; VCode:
+;   pushq   %rbp
+;   movq    %rsp, %rbp
+; block0:
+;   vmovsd  0(%rdi), %xmm3
+;   vmovsd  %xmm0, %xmm3, %xmm0
```
Comment (abrown) [inline cranelift/filetests/filetests/isa/x64/insertlane-avx.clif:67]: Looking at the other cases, I started to worry that one of these instructions actually zeroes bits we do not want to zero. I think it is this case; from the `MOVSD` documentation:

> Legacy version: When the source and destination operands are XMM registers, bits MAXVL:64 of the destination operand remains unchanged. When the source operand is a memory location and destination operand is an XMM registers, the quadword at bits 127:64 of the destination operand is cleared to all 0s, bits MAXVL:128 of the destination operand remains unchanged.

> VEX and EVEX encoded register-register syntax: Moves a scalar double precision floating-point value from the second source operand (the third operand) to the low quadword element of the destination operand (the first operand). Bits 127:64 of the destination operand are copied from the first source operand (the second operand). Bits (MAXVL-1:128) of the corresponding destination register are zeroed

So it is the legacy SSE case where a `movsd 0(%rdi) ....` would zero too many bits but actually in the AVX code we could merge these two `vmovsd` into one. Is there a way to do that?

File: cranelift/filetests/filetests/isa/x64/insertlane-avx.clif [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  v3 = insertlane v0, v2, 1
+  return v3
+}
+
+; VCode:
+;   pushq   %rbp
+;   movq    %rsp, %rbp
+; block0:
+;   movzbq  0(%rdi), %rdx
+;   vpinsrb $1 %xmm0, %rdx, %xmm0
```
Comment (abrown) [inline cranelift/filetests/filetests/isa/x64/insertlane-avx.clif:95]: Missing comma after `$1`?

File: cranelift/filetests/filetests/isa/x64/insertlane.clif [hunk 5]
```diff
@@ -63,8 +63,8 @@ block0(v0: f64x2, v1: i64):
 ;   pushq   %rbp
 ;   movq    %rsp, %rbp
 ; block0:
-;   vmovsd  0(%rdi), %xmm3
-;   vmovsd  %xmm0, %xmm3, %xmm0
+;   movsd   0(%rdi), %xmm3
+;   movsd   %xmm0, %xmm3, %xmm0
```
Comment (abrown) [inline cranelift/filetests/filetests/isa/x64/insertlane.clif:67]: 👍

**Review by abrown — APPROVED** [review]
LGTM but see a couple comments below.

File: cranelift/codegen/src/isa/x64/lower.isle (hunk 1 above)
Comment (alexcrichton) [reply cranelift/codegen/src/isa/x64/lower.isle:2674]: I think [this is the block](https://github.com/bytecodealliance/wasmtime/blob/9ed441e6575685e3c795ab4a504e7f90be01ce3e/cranelift/codegen/src/verifier/mod.rs#L1663-L1687) in the verifier which validates that lanes are always inbounds, so I think we should be good on that front. I can add some extra asserts though to the backend too.

File: cranelift/filetests/filetests/isa/x64/extractlane-avx.clif (hunk 2 above)
Comment (alexcrichton) [reply cranelift/filetests/filetests/isa/x64/extractlane-avx.clif:139]: Indeed!

File: cranelift/filetests/filetests/isa/x64/insertlane-avx.clif (hunk 3 above)
Comment (alexcrichton) [reply cranelift/filetests/filetests/isa/x64/insertlane-avx.clif:67]: Going by [this](https://www.felixcloutier.com/x86/movsd) which I realize isn't official but has been what I've been using so far, my read is that `VEX.128.F2.0F 11 /r: VMOVSD xmm1, xmm2, xmm3` has different semantics than `VEX.128.F2.0F 10 /r: VMOVSD xmm1, m64` so I think that AVX and SSE match here where if you use the register-to-register form it preserves the upper bits but if you use the memory-to-register form it always zeros the upper bits, so I don't think we can fuse?

I'll admit though that this is all real subtle and if the official docs are different I wouldn't be too surprised.

**Comment by alexcrichton** [comment]
I should also note that `pinsrb` and `pinsrw` do not actually perform load sinking at this time due to [this check](https://github.com/bytecodealliance/wasmtime/blob/9ed441e6575685e3c795ab4a504e7f90be01ce3e/cranelift/codegen/src/isa/x64/lower.rs#L85-L93) which is correct for most general-purpose instructions but isn't applicable for `pinsrb` and `pinsrw` specifically as they read the correct byte size. I wasn't sure how to plumb that through though easily to get the load sinking to happen, so the tests reflect how the load isn't sunk currently.

**Pull request merged**